Heroku server

tommcgee opened the pull request:
Buffered Reader now has a timeout, instead of checking for BufferedReader.ready()
As a result the heroku deployment works as intended!

Commit:
Error handling for a null request

Commit:
Works With Heroku

Buffered Reader now has a timeout, instead of checking for BufferedReader.ready()
As a result the heroku deployment works as intended!

cdemyanovich commented on src/main/java/com/server/ServerRunner.java:
```diff
@@ -23,29 +23,32 @@
+            ServerSocket serverSocket = new ServerSocket(serverPort);
+            while (running) {
                 Socket clientSocket = serverSocket.accept();
+                clientSocket.setSoTimeout(100);
```
Why does this solve the problem that you were experiencing?

cdemyanovich commented on src/main/java/com/server/ServerRunner.java:
```diff
@@ -23,29 +23,32 @@
+                request.append((char) mychar);
+            }
+        } catch(Exception e) {
+            return request.toString();
```
This will catch any subclass of `Exception`. Is there a way to be more specific? Why duplicate the `return` statement in the `catch` block?

cdemyanovich commented on src/main/java/com/server/ServerRunner.java:
```diff
@@ -65,13 +68,12 @@ private void respondToRequest (DataOutputStream out, BufferedReader in) throws I
                 response = getResponse(controllerAction);
-            } catch (RouterException e) {
-                System.err.println(e.getMessage());
+            } catch (Exception e) {
```
Why change to a less specific exception?

cdemyanovich commented on src/main/java/com/server/request/Request.java:
```diff
@@ -14,8 +14,8 @@
-        this.request = fullRequestFromServer;
-        header = getRequestHeader();
+            this.request = fullRequestFromServer;
+            header = getRequestHeader();
```
Why the indentation change?

tcmcgee commented:
Hey Craig, I'll address your comments one at a time.


> Why does this solve the problem that you were experiencing?
 `clientSocket.setSoTimeout(100);`

Basically the problem ended up having to do with Heroku not reading the request as we anticipated, particularly the `in.ready()` aspect of the original code was saying the `BufferedReader` was empty when it wasn't. Looking for another solution to this I tried `BufferedReader.readLine()` to get the information, because when there's nothing left to be read it returns an emptyString or null (I can't remember which) but it doesn't work correctly with reading the `\r\n` in the request. So I needed to use `BufferedReader.read()` along with not using `BufferedReader.ready()` to check if it's done reading, the only solution I found to this since `read()` doesn't return anything when there's nothing to read and just hangs, was to set the timeout.

>This will catch any subclass of Exception. Is there a way to be more specific? Why duplicate the return statement in the catch block?

I believe I can change the `Exception` to a `SocketTimeoutException`, and use an empty catch block! Thanks for the feedback here!

> Why change to a less specific exception?
```
-            } catch (RouterException e) {
-                System.err.println(e.getMessage());
+            } catch (Exception e) {
```

This change is in place around the whole thing to stop heroku from failing to a crash and instead if it receives a variety of unexpected requests (that would produce unexpected errors), it would just forget about it and the server would remain running. Do you think we should track down all of the possible exceptions and list them here?

>Why the indentation change?
```
-        this.request = fullRequestFromServer;
-        header = getRequestHeader();
+            this.request = fullRequestFromServer;
+            header = getRequestHeader();
```

Whoops! I think I initially had some debugging around this that was later removed, but the indentation wasn't fixed. Thanks for pointing this out and I'll definitely fix it.

Commit:
Correct indentation and better specify exception when timing out.

- Debugging produced an extra tab of indentation, which is now removed
- Instead of Catching a General `Exception` on timeout, now catching a `SocketTimeOutException`
- removed an extraneous return when reading the full request

cdemyanovich commented:
> and use an empty catch block! 

@tcmcgee Being excited about using an empty `catch` block will extend your apprenticeship at least a week. But, since this is your first offense, perhaps some leniency is appropriate. I'll just scold you:  NEVER WRITE EMPTY `catch` BLOCKS! :-)

cdemyanovich commented:
> Do you think we should track down all of the possible exceptions and list them here?

@tcmcgee and @beccanelson Yes, and shouldn't that be easy in Java, since the documentation and/or the compiler will yell at you if you omit `catch` blocks?

Another good strategy is to catch these low-level exceptions and translate them into one that is appropriate for your current level of abstraction and domain. Any of us mentors could elaborate on that in person if that would help.
